DSpace/dspace-angular
# User agreement

## References
- Fixes #736
- Depends on #856 


## Description
This PR adds a User Agreement at `/info/end-user-agreement` that all registered users have to agree with before being able to do anything else.

This PR also adds the privacy statement page. You can find it at `/info/privacy`. It isn’t referenced anywhere yet in this PR. That is done by #861, which adds a footer link and mentions it in the cookie popup. That PR also adds a footer link to the end user agreement.

## Instructions for Reviewers
Changes made:
- `EndUserAgreementService`: A service to check or update whether or not a user has accepted the agreement. There's currently two ways an agreement can be accepted:
    - A cookie in the user's browser storing a boolean
    - Logged in e-person's `dspace.agreements.end-user` metadata
- Two new guards: `EndUserAgreementCookieGuard` returning true when the cookie is accepted and `EndUserAgreementCurrentUserGuard` returning true when the user's metadata contains "true" OR no user is logged in (anonymous). In case these guards consider the agreement not accepted, it'll return a UrlTree to a new `EndUserAgreementComponent` with the original destination as a redirect url in the query parameters.
- `EndUserAgreementCookieGuard` is configured at route `/register/:token`. This is so the User Agreement can be accepted during registration.
- `EndUserAgreementCurrentUserGuard` is configured on most routes (with the exception of `/reload`, `/register`, `/login`, `/logout`, `/info` (new) `/unauthorized` and `**`)
- `EndUserAgreementComponent`: A component displaying the User Agreement (currently Lorem Ipsum) with a checkbox at the bottom to accept the agreement. This checkbox is automatically selected when: The cookie contains "true" OR the current user's metadata contains "true". When the checkbox is selected, you're able to click "Save". Clicking "Save" will set the cookie (when anonymous) or metadata (when authenticated) and redirect the user back to their original destination.

How to un-agree during testing:
- You can remove the cookie by using your browser's dev tools. 
- You can reset the metadata by saving with an unchecked "agree" checkbox. The save button will be disabled if you uncheck, but you can remove the disabled attribute from the save button in the dev tools, and save anyway to reset the metadata for a user that has already agreed.


Ensure both the cookie and user metadata are not set to "true", before attempting any of these scenarios:
- Visit any route blocked by authentication and login to an authorised account after being redirected to `/login`. You should be redirected to the User Agreement page. Accepting the User Agreement should set the user's metadata to "true" and redirect you back to your original destination.
- Visit any public page and login using the dropdown at the top. Again, you should be redirected to the User Agreement page and after accepting, redirected back to the public page you were on.
- Register a new user. After clicking the link in your email, the User Agreement should pop up. Accepting it should set the cookie to "true" and redirect you back to `/register/:token` to continue the registration.

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_
